### ISSUE TITLE
[Feature request] Batch linear algebra operators

### ISSUE THREAD

**Issue by gpleiss** [pytorch/pytorch#7500]
It would be awesome if there were batched versions of `potrf`, `potrs`, `potri`, and `qr`. The batch Cholesky functions would be useful in [`MultivariateNormal`](https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py#L41), which currently uses a for loop to compute a batch of Cholesky decompositions.

It looks like Magma has batched versions of these functions. Lapack does not, but the Lapack version of this function could always use a for-loop internally.

Open to a PR?
/cc @vishwakftw @SsnL @jacobrgardner

Edit: and batched matrix inverse
Edit [vishwakftw]: Added a checklist

- [x] `solve` a.k.a. `gesv` (#6100)
- [x] `inverse` (#9949)
- [x] `cholesky_solve` a.k.a `potrs` (#13453)
- [x] `cholesky` a.k.a `potrf` (#14017)
- [x] `lu` a.k.a `btrifact` (#14964)
- [x] `triangular_solve` a.k.a `trtrs` (#18025)
- [x] `qr` (#20689)
- [ ] `lstsq` a.k.a `gels`
- [x] `lu_solve` a.k.a `btrisolve` (#24333)
- [x] `svd` (#21588)
- [ ] `cholesky_inverse` a.k.a `potri` (#31275)
- [x] `det` / `logdet` / `slogdet` (#22909)
- [x] `symeig` (#21858)
- [x] `pinverse` (#26095)

**Comment by zou3519** [comment]
Definitely open to a PR.  See [gesv](https://github.com/pytorch/pytorch/pull/6100/files) as an example of a non-batched operation that we added batch support for.

**Comment by vadimkantorov** [comment]
Let's add batched matrix `inverse()` to the top post to not forget it? @soumith

**Comment by fmassa** [comment]
There is a PR open for batches inverse, right? @vadimkantorov could you maybe have a look and review it?

**Comment by fmassa** [comment]
I was referring to https://github.com/pytorch/pytorch/pull/9102
It actually requires some work left, but might be a good starting point

**Comment by vishwakftw** [comment]
@fmassa I have got one completed. I was unaware of #9102 unfortunately.

**Comment by samuela** [comment]
This issue is huge for me. My life would be way easier if everything just had batching by default. And ideally for arbitrary batch shapes, as in `torch.potrf` accepts [2, 3, 5, 5] just as well as [6, 5, 5].

**Comment by vishwakftw** [comment]
@vidhishanair Batched Inverse is available in master now.

**Comment by vidhishanair** [comment]
> @vidhishanair Batched Inverse is available in master now.

Thanks @vishwakftw  This greatly helps 👍 I'll use it in my model and update if there's any discrepency!

**Comment by carlosgmartin** [comment]
Can `torch.det` be added to this list?

**Comment by vishwakftw** [comment]
`torch.slogdet` should be able to give you the determinant. However, certain modifications (minor, I think) have to be made to enable batching for `torch.slogdet` and `torch.logdet`. I'll try to get it done in a few days.

**Comment by el3ment** [comment]
By way of update, it looks like @vishwakftw finished batched `potrs` and it was merged into master a few days ago! Thanks!

**Comment by vishwakftw** [comment]
Update: batched cholesky is merged into master.

**Comment by JurijsNazarovs** [comment]
Since torch.det might be updated, I would also like to see torch.logdet.

**Comment by vishwakftw** [comment]
Yes, this should be done once #14964 is merged into master.

**Comment by SergioRAgostinho** [comment]
Request for  `torch.trace`.

**Comment by vishwakftw** [comment]
torch.trace should be simple. A sample implementation is:

```
def trace(A):
    return A.diagonal(dim1=-2, dim2=-1).sum(-1)
```

**Comment by kefirski** [comment]
Hi, @vishwakftw, Is there any ETA on adding batched det/logdet operations?

**Comment by vishwakftw** [comment]
It should be done by the end of the month. The only thing hindering batched det / logdet / slogdet is the backward pass which requires batched SVD.

**Comment by Gsunshine** [comment]
The end of the month! @vishwakftw 
We really need batched det/logdet/slogdet!
thx!

**Comment by vishwakftw** [comment]
Thanks for your interest. An implementation of batch SVD is available in #21858, but it is a work in progress. Once this is done, the rest should follow.

**Comment by vishwakftw** [comment]
While we don't have batched support for det / logdet / slogdet in the master branch just yet, please feel free to use [vishwakftw:batch-svd](https://github.com/vishwakftw/pytorch/tree/batch-svd) and [vishwakftw:batch-det-logdet-slogdet](https://github.com/vishwakftw/pytorch/tree/batch-det-logdet-slogdet). This combination of these two branches should effectively give you tentative support.

**Comment by calincru** [comment]
> While we don't have batched support for det / logdet / slogdet in the master branch just yet, please feel free to use [vishwakftw:batch-svd](https://github.com/vishwakftw/pytorch/tree/batch-svd) and [vishwakftw:batch-det-logdet-slogdet](https://github.com/vishwakftw/pytorch/tree/batch-det-logdet-slogdet). The combination of these two branches should effectively give you tentative support.

Can you estimate when they'll be merged onto master?

**Comment by vishwakftw** [comment]
I am not sure. I hope to have it reviewed and merged before the next release comes out.

**Comment by calincru** [comment]
When is the next release scheduled for?  Sorry if this is mentioned somewhere, I couldn't find it.

**Comment by calincru** [comment]
> torch.trace should be simple. A sample implementation is:
> 
> ```
> def trace(A):
>     return A.diagonal(dim1=-2, dim2=-1).sum(-1)
> ```

Just curious, why not implement it, though?  It's nice that it can be obtained pretty easily with the current features, but it would help readability and UX in general.

**Comment by vishwakftw** [comment]
We have support for batched det / logdet / slogdet in master.

**Comment by yang-song** [comment]
It would be great if batched `pinverse` can be added.

**Comment by vishwakftw** [comment]
@yang-song batched `pinverse` has been added and is available in the nightly releases.

**Comment by Balandat** [comment]
What's the status on batched `cholesky_inverse`? Would be nice to be able to use this.

**Comment by vishwakftw** [comment]
During benchmarking, I saw a regression for small batch sizes. The PR is usable however.

I’m pretty occupied until the first week of February, so I won’t be able to take address that until then unfortunately.

**Comment by Lezcano** [comment]
If I'm not mistaken, we currently support batches in all the linalg functions, so this one can go 🥳
